drivers/sensor/lis2dh/lis2dh_trigger.c

```diff
@@ -385,11 +381,9 @@ int lis2dh_init_interrupt(const struct device *dev)
 	 */
 
 	/* setup data ready gpio interrupt */
-	lis2dh->gpio_int1 = device_get_binding(cfg->irq1_dev_name);
-	if (lis2dh->gpio_int1 == NULL) {
-		LOG_INF("Cannot get pointer to irq1_dev_name");
-		status = 0;
-		goto end;
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
```

Comment:
Is this right? It seems like the following could happen:

- lis2dh_init(dev) is called
- CONFIG_LIS2DH_TRIGGER=y, so lis2dh_init_interrupt(dev) is called
- dev has no drdy GPIO set up in devicetree, so `cfg->gpio_drdy.port` is NULL
- device_is_ready() returns 0 for NULL argument, so lis2dh_init_interrupt() returns -EINVAL
- lis2dh_init(dev) fails

I'm wondering if the drdy GPIO is not defined for just one instance, even though CONFIG_LIS2DH_TRIGGER=y globally, we should fail the entire device init or not.

Reply:
So, I would go in this way if you agree...

Reply:
Actually I think that testing gpio_drdy.port would be enough. That's because I'm not sure it is possible to configure only gpio_int in DT. I mean, it would be perfectly legit from sensor point of view, but from DT standpoint IDX=1 is valid only if IDX=0 is present. Isn't it?

I tried:
'''
irq-gpios = ,  <&arduino_header 4 GPIO_ACTIVE_HIGH>;
'''

:D

Reply:
>  from DT standpoint IDX=1 is valid only if IDX=0 is present. Isn't it?

The secret sauce is `irq-gpios = <0>, <&arduino_header 4 GPIO_ACTIVE_HIGH>;`

Reply:
> To be honest I think that a flag would be more clear, but I understand also your point to be smaller in size.

Could use a static inline `is_trig_enabled` or so for readability.

Reply:
> The secret sauce is 

I learned this trick here: https://elixir.bootlin.com/linux/latest/source/include/linux/of_gpio.h#L92

We started supporting it in edtlib in 38ede5aded05964054e1706058159bef94bf2d7d though.

Reply:
In this case I should test both.

Reply:
Test both what, sorry? You can test each index independently like you're doing now if I understand your purpose correctly.

Reply:
My last code is like:
```
if (cfg->gpio_drdy.port != NULL || cfg->gpio_int.port != NULL) {
    status = lis2dh_init_interrupt(dev);
    if (status < 0) {
        LOG_ERR("Failed to initialize interrupts.");
        return status;
    }
}
```

Reply:
OK, I see what you mean by "test both": since lis2dh_init_interrupt() is responsible for initializing both interrupts, so you need to test if either of these pointers is not NULL.

But I'm not sure I follow the logic in that function anymore.

I think there's a difference between these 2 things:

1. `(cfg->gpio_drdy.port != NULL) && (device_is_ready(cfg->gpio_drdy.port) == false))`
2. `cfg->gpio_drdy.port == NULL`

I think that 1. should be an error but 2. is not an error.

But 1. can still happen inside lis2dh_init_interrupt if `cfg->gpio_drdy.port != NULL && cfg->gpio_int.port != NULL`, right?

And lis2dh_interrupt still returns zero, which seems wrong, since 1. should be an error.

Am I missing something?

Reply:
So, both means gpio_drdy or gpio_int is configured.

Reply:
> OK, I see what you mean by "test both": since lis2dh_init_interrupt() is responsible for initializing both interrupts, you need to test if either of these pointers is not NULL.
> 
> But I'm not sure I follow the logic in that function anymore.
> 
> I think there's a difference between these 2 things:
> 
> 1. `(cfg->gpio_drdy.port != NULL) && (device_is_ready(cfg->gpio_drdy.port) == false))`
> 2. `cfg->gpio_drdy.port == NULL`
> 
> I think that 1. should be an error but 2. is not an error.
> 
> But 1. can still happen inside lis2dh_init_interrupt if `cfg->gpio_drdy.port != NULL && cfg->gpio_int.port != NULL`, right?
> 
> And lis2dh_interrupt still returns zero, which seems wrong, since 1. should be an error.
> 
> Am I missing something?

What you are saying is correct. It may happen that device is not ready even if gpio_drdy.port is valid. In this case code is not returning an error even if it should. I have similar issue in all sensor drivers. I think that I need an extra test if device_is_ready return false, as the API was changed in #35268. Actually everybody using this API should...

Reply:
I'll resubmit the PR

Reply:
> Actually everybody using this API should...

Well, before, nobody should have been passing NULL to device_is_ready, so existing code which did before the API changed was just wrong, so I guess it's at least not a new issue if it's still wrong.

>  I'll resubmit the PR

Sounds good, I've unresolved this comment in the meantime.